crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3362,6 +3370,38 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn fstring_expression() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            x = 0
+            y = str()
+            z = False
+
+            a = f'hello'
+            b = f'h {x}'
+            c = 'one ' f'single ' f'literal'
+            d = 'first ' f'second({b})' f' third'
+            e = f'-{y}-'
+            f = f'-{y}-' f'--' '--'
+            g = f'{z} == {False} is {True}'
```

Comment:
Let's also add a test with a known value with a format spec and test (for now) that we fallback to `LiteralString`.